message/infrastructure/sql/publisher.go

```diff
@@ -0,0 +1,123 @@
+package sql
+
+import (
+	"database/sql"
+	"sync"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/pkg/errors"
+)
+
+var (
+	ErrPublisherClosed = errors.New("publisher is closed")
+)
+
+type PublisherConfig struct {
+	// SchemaAdapter provides the schema-dependent queries and arguments for them, based on topic/message etc.
+	SchemaAdapter SchemaAdapter
+
+	// MessagesTable is the name of the table that will store the published messages. Defaults to `messages`.
+	MessagesTable string
+}
+
+func (c PublisherConfig) validate() error {
+	if c.SchemaAdapter == nil {
+		return errors.New("schema adapter is nil")
+	}
+
+	return nil
+}
+
+func (c *PublisherConfig) setDefaults() {
+	if c.MessagesTable == "" {
+		c.MessagesTable = "messages"
+	}
+}
+
+// db is implemented both by *sql.DB and *sql.Tx
+type db interface {
+	Prepare(q string) (*sql.Stmt, error)
+}
+
+// Publisher inserts the Messages as rows into a SQL table..
+type Publisher struct {
+	conf PublisherConfig
+
+	db db
+
+	publishWg *sync.WaitGroup
+	closeCh   chan struct{}
+	closed    bool
+}
+
+func NewPublisher(db db, conf PublisherConfig) (*Publisher, error) {
+	conf.setDefaults()
+	if err := conf.validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid config")
+	}
+
+	if db == nil {
+		return nil, errors.New("db is nil")
+	}
+
+	return &Publisher{
+		conf: conf,
+		db:   db,
+
+		publishWg: new(sync.WaitGroup),
+		closeCh:   make(chan struct{}),
+		closed:    false,
+	}, nil
+}
+
+// Publish inserts the messages as rows into the MessagesTable.
```

Comment:
maybe some info that transaction should be handled by the user of Publish (and examples in docs?)

Reply:
32d62eec88fe9c98f7c4bbfb756b08e2362ac812